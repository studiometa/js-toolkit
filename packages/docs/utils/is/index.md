# Value utils

Quickly test values.

- [isDev](./isDev.html)
- [isFunction](./isFunction.html)
- [isDefined](./isDefined.html)
- [isString](./isString.html)
- [isObject](./isObject.html)
- [isNumber](./isNumber.html)
- [isBoolean](./isBoolean.html)
- [isArray](./isArray.html)
