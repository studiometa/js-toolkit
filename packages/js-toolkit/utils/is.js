// eslint-disable-next-line no-undef
export const isDev = typeof __DEV__ !== 'undefined' && __DEV__;

/**
 * Get the type of value.
 * @param   {unknown} value
 * @returns {string}
 */
function type(value) {
  return typeof value;
}

/**
 * Test if the given value is a function.
 *
 * @param   {unknown} value
 * @returns {fn is Function}
 */
export function isFunction(value) {
  return type(value) === 'function';
}

/**
 * Test if a value is defined or not.
 *
 * @param   {unknown} value
 * @returns {boolean}
 */
export function isDefined(value) {
  return type(value) !== 'undefined';
}

/**
 * Test if value is a string.
 * @param   {unknown}  value
 * @returns {value is string}
 */
export function isString(value) {
  return type(value) === 'string';
}

/**
 * Test if the given value is an object.
 *
 * @param {unknown} value
 * @returns {boolean}
 */
export function isObject(value) {
  return type(value) === 'object' && !!value && value.toString() === '[object Object]';
}

/**
 * Test if a given value is a number.
 *
 * @param   {unknown}  value
 * @returns {value is number}
 */
export function isNumber(value) {
  return type(value) === 'number';
}

/**
 * Test if a given value is a boolean.
 * @param   {unknown}  value
 * @returns {value is boolean}
 */
export function isBoolean(value) {
  return type(value) === 'boolean';
}

export const isArray = Array.isArray;
