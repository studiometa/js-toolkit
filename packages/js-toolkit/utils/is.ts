// eslint-disable-next-line no-undef
export const isDev = typeof __DEV__ !== 'undefined' && __DEV__;

/**
 * Test if the given value is a function.
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export const isFunction = (value: unknown): value is Function => typeof value === 'function';

/**
 * Test if a value is defined or not.
 */
export const isDefined = (value: unknown): boolean => typeof value !== 'undefined';

/**
 * Test if value is a string.
 */
export const isString = (value: unknown): value is string => typeof value === 'string';

/**
 * Test if the given value is an object.
 */
export const isObject = (value: unknown): boolean =>
  typeof value === 'object' && !!value && value.toString() === '[object Object]';

/**
 * Test if a given value is a number.
 */
export const isNumber = (value: unknown): value is number => typeof value === 'number' && !Number.isNaN(value);

/**
 * Test if a given value is a boolean.
 */
export const isBoolean = (value: unknown): value is boolean => typeof value === 'boolean';

// eslint-disable-next-line prefer-destructuring
export const isArray = Array.isArray;
