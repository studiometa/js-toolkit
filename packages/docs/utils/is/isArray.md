# isArray

Test if a value is an array.

## Usage

```js
import { isArray } from '@studiometa/js-toolkit/utils';

isArray([]); // true
isArray(1); // false
```

### Parameters

- `value` (`any`): the value to test

### Return value

Returns `true` if the given value is an array, `false` otherwise.
