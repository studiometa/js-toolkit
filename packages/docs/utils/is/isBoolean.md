# isBoolean

Test if a value is a boolean.

## Usage

```js
import { isBoolean } from '@studiometa/js-toolkit/utils';

isBoolean(true); // true
isBoolean('foo'); // false
```

### Parameters

- `value` (`any`): the value to test

### Return value

Returns `true` if the given value is a boolean, `false` otherwise.
