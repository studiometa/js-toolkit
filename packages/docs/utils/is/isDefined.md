# isDefined

Test if a value is defined.

## Usage

```js
import { isDefined } from '@studiometa/js-toolkit/utils';

isDefined('foo'); // true
isDefined(123); // true
isDefined(null); // true
isDefined(); // false
```

### Parameters

- `value` (`any`): the value to test

### Return value

Returns `true` if the given value is not `undefined`, `false` otherwise.
