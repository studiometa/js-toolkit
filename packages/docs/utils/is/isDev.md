# isDev

A constant which is `true` in dev mode (if a global `__DEV__` variable is set and truthy).

## Usage

```js twoslash
import { isDev } from '@studiometa/js-toolkit/utils';

if (isDev) {
  console.log('We are in a development environment.');
}
```
