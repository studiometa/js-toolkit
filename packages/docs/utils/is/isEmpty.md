# isEmpty

Test if a value is empty.

## Usage

```js
import { isEmpty } from '@studiometa/js-toolkit/utils';

isEmpty(); // true
isEmpty(null); // true
isEmpty(''); // true
isEmpty([]); // true
isEmpty({}); // true

isEmpty(0); // false
isEmpty(1); // false
isEmpty([1, 2]); // false
isEmpty({ foo: 'foo' }); // false

class Foo {}
isEmpty(new Foo()); // false
```

### Parameters

- `value` (`any`): the value to test

### Return value

Returns `true` if the given value is considered empty, `false` otherwise.
