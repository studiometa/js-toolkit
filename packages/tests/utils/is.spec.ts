import { describe, it, expect } from 'vitest';
import * as is from '#private/utils/is.js';

const types = {
  boolean: true,
  string: 'string',
  emptyString: '',
  number: 0,
  fn: () => {},
  array: [],
  object: {},
  undefined: undefined,
  null: null,
};
const fns: Array<[string, string | string[]]> = [
  ['isArray', 'array'],
  ['isBoolean', 'boolean'],
  ['isDefined', ['boolean', 'string', 'emptyString', 'number', 'fn', 'array', 'object', 'null']],
  ['isEmptyString', 'emptyString'],
  ['isFunction', 'fn'],
  ['isNull', 'null'],
  ['isNumber', 'number'],
  ['isObject', 'object'],
  ['isString', ['string', 'emptyString']],
];

for (const [name, trueType] of fns) {
  describe(`The "${name}" utility function`, () => {
    it(`should be true with value of type "${trueType}"`, () => {
      for (const [type, value] of Object.entries(types)) {
        expect(is[name](value)).toBe(trueType.includes(type));
      }
    });
    it(`should be false with value different from type "${trueType}"`, () => {
      for (const [type, value] of Object.entries(types)) {
        expect(is[name](value)).toBe(trueType.includes(type));
      }
    });
  });
}

describe('The "isEmpty" utility function', () => {
  it('should return true when the given value is empty', () => {
    expect(is.isEmpty()).toBe(true);
    expect(is.isEmpty('')).toBe(true);
    expect(is.isEmpty(null)).toBe(true);
    expect(is.isEmpty([])).toBe(true);
    expect(is.isEmpty({})).toBe(true);
  });

  it('should return false when the given value is not empty', () => {
    expect(is.isEmpty(Number.NaN)).toBe(false);
    expect(is.isEmpty(1)).toBe(false);
    expect(is.isEmpty(0)).toBe(false);
    expect(is.isEmpty('foo')).toBe(false);
    expect(is.isEmpty(true)).toBe(false);
    expect(is.isEmpty(false)).toBe(false);
    expect(is.isEmpty([1, 2])).toBe(false);
    expect(is.isEmpty({ foo: 'foo' })).toBe(false);
    expect(is.isEmpty(/regex/)).toBe(false);

    class Foo {}
    const foo = new Foo();
    expect(is.isEmpty(foo)).toBe(false);
  });
});

describe('The "isNumber" utility function', () => {
  it('should return false for NaN', () => {
    expect(is.isNumber(Number.NaN)).toBe(false);
    expect(is.isNumber(Number('100px'))).toBe(false);
  });
});
