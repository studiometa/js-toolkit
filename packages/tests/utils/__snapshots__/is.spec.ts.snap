// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The "isArray" utility function should work with value of type "array" 1`] = `true`;

exports[`The "isArray" utility function should work with value of type "boolean" 1`] = `false`;

exports[`The "isArray" utility function should work with value of type "fn" 1`] = `false`;

exports[`The "isArray" utility function should work with value of type "number" 1`] = `false`;

exports[`The "isArray" utility function should work with value of type "object" 1`] = `false`;

exports[`The "isArray" utility function should work with value of type "string" 1`] = `false`;

exports[`The "isArray" utility function should work with value of type "undefined" 1`] = `false`;

exports[`The "isBoolean" utility function should work with value of type "array" 1`] = `false`;

exports[`The "isBoolean" utility function should work with value of type "boolean" 1`] = `true`;

exports[`The "isBoolean" utility function should work with value of type "fn" 1`] = `false`;

exports[`The "isBoolean" utility function should work with value of type "number" 1`] = `false`;

exports[`The "isBoolean" utility function should work with value of type "object" 1`] = `false`;

exports[`The "isBoolean" utility function should work with value of type "string" 1`] = `false`;

exports[`The "isBoolean" utility function should work with value of type "undefined" 1`] = `false`;

exports[`The "isDefined" utility function should work with value of type "array" 1`] = `true`;

exports[`The "isDefined" utility function should work with value of type "boolean" 1`] = `true`;

exports[`The "isDefined" utility function should work with value of type "fn" 1`] = `true`;

exports[`The "isDefined" utility function should work with value of type "number" 1`] = `true`;

exports[`The "isDefined" utility function should work with value of type "object" 1`] = `true`;

exports[`The "isDefined" utility function should work with value of type "string" 1`] = `true`;

exports[`The "isDefined" utility function should work with value of type "undefined" 1`] = `false`;

exports[`The "isFunction" utility function should work with value of type "array" 1`] = `false`;

exports[`The "isFunction" utility function should work with value of type "boolean" 1`] = `false`;

exports[`The "isFunction" utility function should work with value of type "fn" 1`] = `true`;

exports[`The "isFunction" utility function should work with value of type "number" 1`] = `false`;

exports[`The "isFunction" utility function should work with value of type "object" 1`] = `false`;

exports[`The "isFunction" utility function should work with value of type "string" 1`] = `false`;

exports[`The "isFunction" utility function should work with value of type "undefined" 1`] = `false`;

exports[`The "isNumber" utility function should work with value of type "array" 1`] = `false`;

exports[`The "isNumber" utility function should work with value of type "boolean" 1`] = `false`;

exports[`The "isNumber" utility function should work with value of type "fn" 1`] = `false`;

exports[`The "isNumber" utility function should work with value of type "number" 1`] = `true`;

exports[`The "isNumber" utility function should work with value of type "object" 1`] = `false`;

exports[`The "isNumber" utility function should work with value of type "string" 1`] = `false`;

exports[`The "isNumber" utility function should work with value of type "undefined" 1`] = `false`;

exports[`The "isObject" utility function should work with value of type "array" 1`] = `false`;

exports[`The "isObject" utility function should work with value of type "boolean" 1`] = `false`;

exports[`The "isObject" utility function should work with value of type "fn" 1`] = `false`;

exports[`The "isObject" utility function should work with value of type "number" 1`] = `false`;

exports[`The "isObject" utility function should work with value of type "object" 1`] = `true`;

exports[`The "isObject" utility function should work with value of type "string" 1`] = `false`;

exports[`The "isObject" utility function should work with value of type "undefined" 1`] = `false`;

exports[`The "isString" utility function should work with value of type "array" 1`] = `false`;

exports[`The "isString" utility function should work with value of type "boolean" 1`] = `false`;

exports[`The "isString" utility function should work with value of type "fn" 1`] = `false`;

exports[`The "isString" utility function should work with value of type "number" 1`] = `false`;

exports[`The "isString" utility function should work with value of type "object" 1`] = `false`;

exports[`The "isString" utility function should work with value of type "string" 1`] = `true`;

exports[`The "isString" utility function should work with value of type "undefined" 1`] = `false`;
